src/sentry/utils/snuba.py

```diff
@@ -0,0 +1,162 @@
+from __future__ import absolute_import
+
+from dateutil.parser import parse as parse_datetime
+import json
+import requests
+import six
+
+from sentry.models import Group, GroupHash, Environment, Release, ReleaseProject
+from sentry.utils.dates import to_timestamp
+
+SNUBA = 'http://localhost:5000'
+
+
+def query(start, end, groupby, conditions=None, filter_keys=None,
+          aggregation=None, aggregateby=None, rollup=None, arrayjoin=None):
+    """
+    Sends a query to snuba.
+
+    `conditions`: A list of (column, operator, literal) conditions to be passed
+    to the query. Conditions that we know will not have to be translated should
+    be passed this way (eg tag[foo] = bar).
+
+    `filter_keys`: A dictionary of {col: [key, ...]} that will be converted
+    into "col IN (key, ...)" conditions. These are used to restrict the query to
+    known sets of project/issue/environment/release etc. Appropriate
+    translations (eg. from environment model ID to environment name) are
+    performed on the query, and the inverse translation performed on the
+    result. The project_id(s) to restrict the query to will also be
+    automatically inferred from these keys.
+    """
+    groupby = groupby or []
+    conditions = conditions or []
+    filter_keys = filter_keys or {}
+
+    # Forward and reverse translation maps from model ids to snuba keys, per column
+    snuba_map = {col: get_snuba_map(col, keys) for col, keys in six.iteritems(filter_keys)}
+    snuba_map = {k: v for k, v in six.iteritems(snuba_map) if k is not None and v is not None}
+    rev_snuba_map = {col: dict(reversed(i) for i in keys.items())
+                     for col, keys in six.iteritems(snuba_map)}
+
+    for col, keys in six.iteritems(filter_keys):
+        if col in snuba_map:
+            keys = [snuba_map[col][k] for k in keys]
+        conditions.append((col, 'IN', keys))
+
+    # project_ids will be the set of projects either referenced directly as
+    # passed-in keys for project_id, or indrectly (eg the set of projects
+    # related to the queried set of issues or releases)
+    project_ids = [get_project_ids(k, ids) for k, ids in six.iteritems(filter_keys)]
+    if all(not ids for ids in project_ids):
+        raise Exception("No project_id filter, or none could be inferred from other filters.")
+    project_ids = list(set.intersection(*[set(ids) for ids in project_ids if ids]))
```

Comment:
Is it `intersection` and not union? If it doesn't go with the largest possible selection of `project_id`s then why not just have the user pass in `project_id`s themselves? I guess I'm not sure why we don't require them as a top level param anyway?

Reply:
for something like `get_group_tag_value_count(project_id, group_id, environment_id)` in tagstore; if you passed in project_id = 1, and an environment_id from project 2, and a group_id from project 3 then:
1. I guess you have a bug anyway, because why are you sending these unrelated things?
2. Its pointless to do a query across the union of all 3 projects, (assuming all conditions are `AND`ed together) because there will be no matches.

Reply:
Oh, to answer the other question. We don't use user-passed in project_ids exclusively because sometimes they don't exist. eg with `TSDBModel.frequent_releases_by_group` and others, we only have the group_id, environment_id etc. 

So the main purpose of this whole block is just to infer the project_id for snuba purposes, when we are not passed a project id. In the cases where we _are_ passed a project id, it doesn't really do much yeah,

Reply:
OK, I definitely buy it not being required. But I wonder if it should still be either,

1. toplevel and not required, but if provided it is strictly used as the `project_id` filter
2. keep what you have an assert that the projects match, because I think you nailed my confusion that it just feels weird to magically pull out `project_id`s and some of them vanish via `intersection`